File: admitad/brd/get_product.py
import re
import random
import requests
from bs4 import BeautifulSoup
from get_sizes import get_sizes
from get_images import get_images
from get_categories import get_categories
from get_transliterate import get_transliterate
from get_info import get_info


def get_product(html, link, gender):
    try:
        soup = BeautifulSoup(html, 'lxml')
        if gender == 'men': gender = 'Мужской'
        else: gender = 'Женский'
        sizes_proto = soup.find_all('label', class_=re.compile('options-item'))
        sizes = []
        for size in sizes_proto:
            if 'disabled' not in size.get('class'):
                result = size.find('span').text.strip()
                if len(result) == 4 and 'XX' not in result: result = result[0:2] + '/' + result[2:4]
                sizes.append(result.replace(',', '.'))
        if len(sizes) == 0: return
        if sizes[0] == '': return
        name = soup.find('h1').text
        price = int(soup.find('span', { 'itemprop': 'price' }).text)
        oldprice = int(soup.find('span', class_='price_obsolete').text.replace(' Р', ''))
        sale = int(soup.find('div', class_=re.compile('product-card__label--sale')).text.replace('-', '').replace('%', ''))
        brand = soup.find('a', class_='brand').find('img').get('alt')
        color = soup.find('div', class_='options').find('span').text.replace('ё', 'е').lower()
        try: desc = soup.find('div', class_='info-description').text.strip()
        except: desc = False
        info = []
        for property in soup.find('div', class_='details-description').find_all('li'):
            info.append(property.text.strip())
        images = []
        for img in soup.find('section', class_='desktop_images').find_all('a', class_=re.compile('cloud-zoom')):
            images.append(img.find('img').get('src'))
        breadcrumbs_proto = soup.find('nav', class_='breadcrumbs').find_all('span', { 'itemprop': 'name' })
        breadcrumbs = []
        for crumb in breadcrumbs_proto:
            breadcrumbs.append(crumb.text)
        category = ''
        if 'одежда' in breadcrumbs[2].lower(): category = 'Одежда'
        else:
            if 'обувь' in breadcrumbs[2].lower(): category = 'Обувь'
        if category == '': category = 'Аксессуары'
        subcategory = breadcrumbs[3]\
            .replace(' мужская', '').replace('Мужская ', '').replace(' мужские', '').replace('Мужские', '').replace('Женские ', '')\
            .replace('Женская ', '').replace(' женская', '').replace(' женские', '').strip().capitalize()
        return {
            'id': round(random.uniform(1000000000, 9999999999)),
            'age': 'Взрослый',
            'benefit': oldprice - price,
            'brand': brand.upper(),
            'brandCountry': False,
            'brandCountry_t': False,
            'category': category,
            'category_t': get_transliterate(category),
            'color': color,
            'color_t': get_transliterate(color) if color else color,
            'country': False,
            'country_t': False,
            'delivery': ['ru'],
            'deliveryPrice': False,
            'description': desc,
            'gender': gender,
            'name': name,
            'info': info,
            'installment': True,
            'images': images,
            'like': 0,
            'link': link,
            'oldprice': oldprice,
            'pp': 'admitad',
            'price': price,
            'sale': sale,
            'season': False,
            'season_t': False,
            'shop': 'brd',
            'sizes': sizes,
            'style': False,
            'style_t': False,
            'structure': False,
            'subcategory': subcategory,
            'subcategory_t': get_transliterate(subcategory).replace(' ', '-').lower()
        }
    except Exception as ex:
        print(ex)
        print(f'{link} НЕ собран!')
        return

# res = requests.get('https://www.brd.ru/product/adynp03076-rsd0-bryuki-chinos-worker-relaxed?utm_source=admitad&utm_medium=cpa&utm_campaign=1392094&admitad_uid=3911ed74823043c1eec359d596ce0bc9&tagtag_uid=3911ed74823043c1eec359d596ce0bc9')
# print(get_product(res.text, 'link', 'gender'))

File: admitad/brd/getProduct.js
const axios = require('axios')
const { JSDOM } = require('jsdom')
const getOptions = require('./getOptions')
const getTransliterate = require('./getTransliterate')
const convertSizes = require('../../utils/getSizesBrd');
const iconv = require('iconv-lite');


module.exports = async function getProduct (productLink, gender, imageCatalog) {
  try {
    // получаю данные по товару
    const productOptions = getOptions(null, null, productLink)
    const productRes = await axios.post( ...productOptions )
    const ruHtmlProduct = iconv.decode(productRes.data, "win1251")
    const dom = new JSDOM(ruHtmlProduct).window.document
    if (dom.querySelector('div.last-item-list') === null && dom.querySelector('span.last-item-text').textContent === 'Нет в наличии.') throw new Error('Во время парсинга товар перестал быть в наличии');

    const breadCrumbs = Array.from(dom.querySelector('nav[class="breadcrumbs"]').querySelectorAll('span[itemprop="name"]'))
      .map(crumb => crumb.textContent.toLowerCase())

    const imageLinksArray = Array.from(dom.querySelector('section.desktop_images').querySelectorAll('a'))
      .map(link => link.href)

    const fullNameArr = dom.querySelector('header h1[itemprop="name"]').textContent.split(' ')
    // функция выбирает из массива слова на кириллице и без символа "-"
    let nameArrayCyrillic = fullNameArr.reduce((acc, word) => {
      if (/^[А-ЯЁ-]+$/i.test(word)) acc.push(word)
      return acc
    }, [])

    let category;
    let subcategory;
    if (breadCrumbs.length > 2) {
      breadCrumbs[2].indexOf('одежда') !== -1
      ? category = 'Одежда'
      : breadCrumbs[2].indexOf('обувь') !== -1
        ? category = 'Обувь'
        : category = 'Аксессуары';
      subcategory = breadCrumbs[3].toLowerCase().replaceAll('мужской', '').replaceAll('мужская', '').replaceAll('мужские', '')
        .replaceAll('женский', '').replaceAll('женская', '').replaceAll('женские', '').trim()
      subcategory = subcategory[0].toUpperCase() + subcategory.slice(1)
    } else {
      category = 'Одежда'
      subcategory = nameArrayCyrillic.join(' ').toLowerCase().replaceAll('мужской', '').replaceAll('мужская', '').replaceAll('мужские', '')
        .replaceAll('женский', '').replaceAll('женская', '').replaceAll('женские', '').trim()
      subcategory = subcategory[0].toUpperCase() + subcategory.slice(1)
    }
    subcategory = subcategory.replaceAll('  ', ' ').trim()
    if (subcategory === 'Ботинки сноуборд') subcategory = 'Ботинки для сноуборда'
    if (subcategory === 'Ботинки для сноуборда') category = 'Обувь'
    if (subcategory === 'Головные уборы для сноубординга') category = 'Аксессуары'
    if (subcategory === 'Леггинсы') subcategory = 'Легинсы'
    if (subcategory === 'Спортивные леггинсы') subcategory = 'Легинсы'
    if (subcategory === 'Спортивные шорты') subcategory = 'Шорты'
    if (subcategory === 'Сноубордические перчатки') subcategory = 'Сноубордические перчатки и варежки'
    if (subcategory === 'Сноубордические перчатки и варежки') category = 'Аксессуары'
    if (subcategory === 'Флисовая толстовка на молнии') subcategory = 'Флисовые толстовки'
    if (subcategory === 'Флисовая толстовка') subcategory = 'Флисовые толстовки'
    if (subcategory === 'Толстовки, худи, свитшоты') subcategory = 'Толстовки'
    if (subcategory === 'Шлепанцы (сланцы)') subcategory = 'Сланцы'
    if (subcategory === 'Рашгарды для серфинга и плавания') subcategory = 'Рашгарды для плавания и серфинга'
    if (subcategory === 'Бордшорты (шорты для плавания)') subcategory = 'Бордшорты'
    if (subcategory === 'Бордшорты') category = 'Одежда'
    if (subcategory === 'Гидрокостюмы') category = 'Одежда'
    if (subcategory === 'Носки') category = 'Одежда'
    if (subcategory === 'Плавательные шорты') category = 'Одежда'
    if (subcategory === 'Гидрообувь') category = 'Обувь'


    let brand = dom.querySelector('meta[itemprop=brand]').getAttribute('content').toUpperCase()
    nameArrayCyrillic.push(brand)

    let infoDetails = dom.querySelector('div.details').querySelector('meta[itemprop="description"]').getAttribute("content").split('\n')
      .map(infoElement => infoElement.trim())

    let sale = parseInt(dom.querySelector('div.product-card__label.product-card__label--sale').textContent)
    if (sale < 0) sale = sale * -1

    let sizes = [];
    const options = dom.querySelector('div.option.text')
    let labels = Array.from(options.querySelectorAll('.options-item.js-ab-prinfo-add-popup-size-hover'));
    for (let label of labels) {
      if (label.getAttribute('class') === 'options-item js-ab-prinfo-add-popup-size-hover ') sizes.push(label);
    }
    sizes = Array.from(sizes).map(size => size.querySelector('span').textContent.replaceAll('W', '').replaceAll('/', '-').replaceAll(' ', '').replace(/\u00A0/g, "").replaceAll('-L', '-'));

    const oldprice = parseInt(dom.querySelector('div.prices').querySelector('span.price_obsolete').textContent)
    const price = parseInt(dom.querySelector('div.prices').querySelector('span[itemprop="price"]').textContent)

    let structure = dom.querySelector('div.composition-info')
    if (structure) {
      structure = structure.textContent.split('/');
      structure[0] = structure[0].split(':')[1];
      structure.forEach(num => structure[num] = structure[num] ? structure[num].trim() : 'Неизвестно')
    } else structure = false

    if (sizes[0] === '') sizes = [ 'one size' ]
    for (let sizeNum in  sizes) {
      if (sizes[sizeNum] === 'OneSize' || sizes[sizeNum] === '1SZ' || sizes[sizeNum] === 'U') sizes = [ 'one size' ];
      if ((category === 'Обувь') && (sizes[sizeNum].indexOf('H') !== -1 || sizes[sizeNum].indexOf('Н') !== -1)) {
        sizes[sizeNum] = sizes[sizeNum].replaceAll('H', '').replaceAll('Н', '').split('-');
        sizes[sizeNum] = Math.max(...sizes[sizeNum]).toString();
      }
      if ((brand === 'DC SHOES' && subcategory === "Брюки") && (sizes[sizeNum].indexOf('-') !== -1)) sizes[sizeNum] = sizes[sizeNum].split('-')[0];
      if (subcategory === 'Платья') sizes[sizeNum] = sizes[sizeNum].split('-')[0];
      if (brand === 'ELEMENT' && subcategory === "Футболки, поло, лонгсливы") sizes[sizeNum] = sizes[sizeNum].split('-')[0];
      if (category === 'Одежда') sizes[sizeNum] = sizes[sizeNum].replaceAll('XXXX', '4X').replaceAll('XXX', '3X').replaceAll('XX', '2X');
      // skip products with invalid sizes
      if ((parseInt(sizes[sizeNum]) && parseInt(sizes[sizeNum]) > 100)) return
      if (!(/\w+$/i.test(sizes[sizeNum]))) return
    }

    let generalRating;
    const reviewsList = [];
    const hasReviews = dom.querySelector('div.totals')
    if (hasReviews !== null) {
      generalRating = hasReviews.querySelector('b').textContent;
      for (let review of dom.querySelector('div.reviews-list').querySelectorAll('div.review')) {
        if (review.querySelector('meta[itemprop="ratingValue"]') !== null) {
          const reviews = {
            rating: review.querySelector('meta[itemprop="ratingValue"]').getAttribute('content'),
            author: review.querySelector('div.author').querySelector('span[itemprop="name"]').textContent,
            text: review.querySelector('div.review-body').querySelector('p').textContent,
            date: review.querySelector('div.date').textContent
          }
          reviewsList.push(reviews)
        }
      }
    }

    const product = {
      id: Math.floor(Math.random() * 9e9) + 1e9,
      age: 'Взрослый',
      benefit: oldprice - price,
      brand,
      brandCountry: false,
      brandCountry_t: false,
      category,
      category_t: getTransliterate(category),
      color: dom.querySelector('dl.attributes').querySelectorAll('dd')[dom.querySelector('dl.attributes').querySelectorAll('dd').length - 1].textContent,
      color_t: getTransliterate(dom.querySelector('dl.attributes').querySelectorAll('dd')[dom.querySelector('dl.attributes').querySelectorAll('dd').length - 1].textContent),
      country: false,
      country_t: false,
      create: new Date,
      delivery: ['ru','rb','kz','kg','am'],
      deliveryPrice: false,
      description: false,
      gender: gender === 'men'
        ? 'Мужской'
        : 'Женский',
      installment: true,
      imageCatalog,
      images: imageLinksArray,
      like: 0,
      link: productLink,
      name: nameArrayCyrillic.join(' ').replaceAll('Мужской', '').replaceAll('Мужская', '').replaceAll('Мужские', '')
        .replaceAll('Женский', '').replaceAll('Женская', '').replaceAll('Женские', '').trim(),
      shop: 'brd',
      info: infoDetails,
      oldprice: oldprice,
      pp: 'admitad',
      price: price,
      sale: sale,
      season: false,
      season_t: false,
      sizes,
      style: false,
      style_t: false,
      structure: structure,
      subcategory,
      subcategory_t: getTransliterate(subcategory),
      reviews: hasReviews !== null
        ? {
            rating: generalRating,
            reviews: reviewsList
          }
        : false,
      bonus: dom.querySelector('div.payment-variants').querySelector('div.payment-variants-list.credit-card') !== null
        ? {
            saleOnlinePurchases: parseInt((dom.querySelector('div.payment-variants').querySelector('div.payment-variants-list.credit-card').querySelector('span.subtext').textContent).match(/\d+/))
          }
        : false
    }
    product.sizes = convertSizes({ sizes, brand, subcategory, gender } = product)
      .map(size => size.replaceAll(',', '.').replaceAll('XXL', '2XL'))
    if (product.imageCatalog === '') return
    return product
  } catch (e) {
    console.log('Товар', productLink, 'не собран', e)
    return null
  }
}
